### PR TITLE
Ruby 2.6: Add #to_h with block argument to Array, Enum, ENV, Hash and Struct

### PR DESCRIPTION
**Pull request by nomadium**
Hi folks,

This PR implements another Ruby 2.6 feature: #to_h method in Array, Enumerable, ENV, Hash and Struct classes now can map their elements to new keys and values to a given block.

For more information, please see feature #15143 [1].

I believe all the related tests (MRI and ruby/spec) are passing:i [2], [3], [4], [5], [6], [7], [8], [9], [10], [11].

For reference, I include a link to the commit introducing the functionality in MRI. [12]

Thanks in advance for any feedback.

 1. https://bugs.ruby-lang.org/issues/15143
 2. https://github.com/ruby/ruby/blob/v2_6_0_preview3/spec/ruby/core/array/to_h_spec.rb#L38-L41
 3. https://github.com/ruby/ruby/blob/v2_6_0_preview3/spec/ruby/core/enumerable/to_h_spec.rb#L47-L53
 4. https://github.com/ruby/ruby/blob/v2_6_0_preview3/spec/ruby/core/env/to_h_spec.rb#L7-L17
 5. https://github.com/ruby/ruby/blob/v2_6_0_preview3/spec/ruby/core/hash/to_h_spec.rb#L10-L14
 6. https://github.com/ruby/ruby/blob/v2_6_0_preview3/spec/ruby/core/struct/to_h_spec.rb#L16-L22
 7. https://github.com/ruby/ruby/blob/v2_6_0_preview3/test/ruby/test_array.rb#L1634-L1653
 8. https://github.com/ruby/ruby/blob/v2_6_0_preview3/test/ruby/test_enum.rb#L183-L202
 9. https://github.com/ruby/ruby/blob/v2_6_0_preview3/test/ruby/test_env.rb#L400-L404
10. https://github.com/ruby/ruby/blob/v2_6_0_preview3/test/ruby/test_hash.rb#L868-L876
11. https://github.com/ruby/ruby/blob/v2_6_0_preview3/test/ruby/test_struct.rb#L365-L370
12. https://github.com/ruby/ruby/commit/abe75149d14d3286d3051c9e961ab6473a243a19

### CODE REVIEW

File: core/src/main/java/org/jruby/RubyArray.java [hunk 1]
```diff
@@ -2047,12 +2047,17 @@ public IRubyObject to_h(ThreadContext context) {
         RubyHash hash = useSmallHash ? RubyHash.newSmallHash(runtime) : RubyHash.newHash(runtime);
 
         for (int i = 0; i < realLength; i++) {
-            IRubyObject elt = eltInternal(i).checkArrayType();
-            if (elt == context.nil) {
-                throw context.runtime.newTypeError("wrong element type " + eltInternal(i).getMetaClass().getRealClass() + " at " + i + " (expected array)");
+            IRubyObject e = eltInternal(i);
+            IRubyObject elt, key_value_pair;
```
Comment (enebo) [inline core/src/main/java/org/jruby/RubyArray.java]: Could you move decls to their first assignment:

```java
  IRubyObject elt = block.isGiven() // ...
```

File: core/src/main/java/org/jruby/RubyGlobal.java [hunk 2]
```diff
@@ -434,8 +435,9 @@ public IRubyObject to_s() {
         }
 
         @JRubyMethod
-        public RubyHash to_h(){
-            return to_hash();
+        public RubyHash to_h(ThreadContext context, Block block){
```
Comment (kares) [inline core/src/main/java/org/jruby/RubyGlobal.java]: we should consider keeping these public method (on RubyXXX types) signatures around (marked as deprecated)
due exts occasionally calling methods directly e.g. 
`@Deprecated public RubyHash to_h() { return to_h(getCurrentContext(), NULL_BLOCK) }`

File: core/src/main/java/org/jruby/RubyArray.java (hunk 1 above)
Comment (nomadium) [reply core/src/main/java/org/jruby/RubyArray.java]: Fixed.

File: core/src/main/java/org/jruby/RubyGlobal.java (hunk 2 above)
Comment (nomadium) [reply core/src/main/java/org/jruby/RubyGlobal.java]: Thanks @kares, I didn't think about that, good catch. I just updated only this method following your recommendation, but just to clarify: do you consider I should to the same backward compatible change for all the public ```to_h``` modified methods in this PR?

Comment (kares) [reply core/src/main/java/org/jruby/RubyGlobal.java]: yeah, that would be the same - to usually deprecate public methods on (core) RubyXXX.java types

**Comment by lopex** [comment]
MRI seems to have refactored rb_hash_set_pair (used in enum, hash, struct) where they have centralized checks for "wrong element type" and "element has wrong array length".
We seem to duplicate a lot.

**Comment by enebo** [comment]
@lopex perhaps a good follow up commit?

**Comment by lopex** [comment]
@enebo yeah I'll do that, hopefuly on this weekend, and have some more things on the table regarding match data, and 2.6 String#split, etc

File: core/src/main/java/org/jruby/RubyGlobal.java (hunk 2 above)
Comment (nomadium) [reply core/src/main/java/org/jruby/RubyGlobal.java]: @kares This should be fixed in the most recent version of this PR.

**Comment by nomadium** [comment]
@enebo @kares I think I addressed your comments. Please take a look at this again when you have a chance.

**Comment by kares** [comment]
LGTM

**Pull request merged**